### PR TITLE
Patch warning note endpoint

### PR DESCRIPTION
**Pull request by LBHBReynolds-Carr**
## Link to JIRA ticket

https://hackney.atlassian.net/browse/SCS-705

### *What is the problem we're trying to solve*

As a user, I want to update Warning Notes in response to a review

### *What changes have we introduced*

- Added PatchWarningNoteRequest
- Added schema for sccv_Warning_note_review table
- Added WarningNoteReview infrastructure object
- Added PatchWarningNote method to gateway with no response
- Created PostNoteReview method to gateway with no response
- Created PatchWarningNote method in gateway
- Created PostWarningNoteReview method in gateway
- PostWarningNoteReview only exists as a method for PatchWarningNote
- Created ExecutePatch method in WarningNoteUseCase
- Created PatchWarningNote method in controller
- Hard coded PostWarningNote method in gateway to set status to open

#### _Checklist_

- [x] Added tests to cover all new production code
- [x] Added comments to the README or updated relevant documentation _(add link to documentation)_, where necessary.
- [x] Checked all code for possible refactoring
- [x] Code pipeline builds correctly

### *Follow up actions after merging PR*

Get PostWarningNoteReview

### CODE REVIEW

**Comment by JohnFarrellDev** [comment]
In future can we keep individual commits shorter and more frequent, commit per each bit of useful functionality such as a single new gateway etc.

File: SocialCareCaseViewerApi/V1/Boundary/Requests/PatchWarningNoteRequest.cs [hunk 1]
```diff
@@ -0,0 +1,20 @@
+using System;
+using System.ComponentModel.DataAnnotations;
+
+namespace SocialCareCaseViewerApi.V1.Boundary.Requests
+{
+    public class PatchWarningNoteRequest
```
Comment (JohnFarrellDev) [inline SocialCareCaseViewerApi/V1/Boundary/Requests/PatchWarningNoteRequest.cs:6]: Can we make use of FluentValidation here similar to `CreateAllocationRequest`?

**Review by JohnFarrellDev — CHANGES_REQUESTED** [review]
I'd be happy to go through this with you tomorrow to help and explain any of these suggested changes.

**Comment by JohnFarrellDev** [comment]
We must add the use case to our Startup.cs, if we also create a validation class it is required in Startup.cs

File: .editorconfig [hunk 2]
```diff
@@ -144,3 +144,5 @@ dotnet_diagnostic.CA2227.severity = none
 # Possible solution by making private and declaring using properties
 dotnet_diagnostic.CA1051.severity = none
 
+# delete before deployment
```
[the remark this answers is not in the document]
Comment (LBHBReynolds-Carr) [reply .editorconfig:147]: Good job I left that comment! Deleted.

File: SocialCareCaseViewerApi/V1/Controllers/SocialCareCaseViewerApiController.cs [hunk 3]
```diff
@@ -407,5 +407,20 @@ public IActionResult GetWarningNote([FromQuery] GetWarningNoteRequest request)
...
+        /// </summary>
+        /// <param name="request"></param>
+        /// <response code="204">Amended successfully</response>
+        /// <response code="500">There was a problem updating the record</response>
+        [ProducesResponseType(StatusCodes.Status204NoContent)]
+        [HttpPatch]
+        [Route("warningnotes")]
+        public IActionResult PatchWarningNote([FromBody] PatchWarningNoteRequest request)
+        {
+            _warningNoteUseCase.ExecutePatch(request);
```
[the remark this answers is not in the document]
Comment (LBHBReynolds-Carr) [reply SocialCareCaseViewerApi/V1/Controllers/SocialCareCaseViewerApiController.cs:422]: I've added a 404 response

File: SocialCareCaseViewerApi/V1/Exceptions/CustomExceptions.cs [hunk 4]
```diff
@@ -42,4 +42,14 @@ public class PostWarningNoteException : Exception
     {
         public PostWarningNoteException(string message) : base(message) { }
     }
+
+    public class PatchWarningNoteException : Exception
+    {
+        public PatchWarningNoteException(string message) : base(message) { }
+    }
+
+    public class PostWarningNoteReviewException : Exception
```
[the remark this answers is not in the document]
Comment (LBHBReynolds-Carr) [reply SocialCareCaseViewerApi/V1/Exceptions/CustomExceptions.cs:51]: Removed

File: terraform/production/.editorconfig [hunk 5]
```diff
@@ -0,0 +1,146 @@
+[*]
```
[the remark this answers is not in the document]
Comment (LBHBReynolds-Carr) [reply terraform/production/.editorconfig:1]: I'm not entirely sure why this entire document is reading as being changed, this matches what's currently in the editor config on the dev branch

File: SocialCareCaseViewerApi.Tests/V1/Controllers/SocialCareCaseViewerApiControllerTests.cs [hunk 6]
```diff
@@ -700,6 +700,16 @@ public void GetWarningNoteReturns404IfDocumentNotDound()
             response.StatusCode.Should().Be(404);
             response.Value.Should().Be("Document Not Found");
         }
+
+        [Test]
```
[the remark this answers is not in the document]
Comment (LBHBReynolds-Carr) [reply SocialCareCaseViewerApi.Tests/V1/Controllers/SocialCareCaseViewerApiControllerTests.cs:704]: Testing for 404 catch added

File: SocialCareCaseViewerApi.Tests/V1/Gateways/DatabaseGateway.Tests.cs [hunk 7]
```diff
@@ -437,15 +437,11 @@ public void CreatingAPersonShouldCreateCorrectRecordsAndAuditRecordsInTheDatabas
         [Test]
         public void PostWarningNoteShouldInsertIntoTheDatabase()
         {
-            Person person = new Person()
-            {
-                FullName = "full name"
-            };
-            DatabaseContext.Persons.Add(person);
-            DatabaseContext.SaveChanges();
+            Person stubbedPerson = CreateStubbedPerson();
```
[the remark this answers is not in the document]
Comment (LBHBReynolds-Carr) [reply SocialCareCaseViewerApi.Tests/V1/Gateways/DatabaseGateway.Tests.cs]: TestHelpers.CreatePerson() now utilised

File: SocialCareCaseViewerApi.Tests/V1/Gateways/DatabaseGateway.Tests.cs [hunk 8]
```diff
@@ -491,15 +487,11 @@ public void PostWarningNoteShouldThrowAnErrorIfThePersonIdIsNotInThePersonTable(
...
         public void PostWarningNoteShouldCallInsertCaseNoteMethod()
         {
-            Person person = new Person()
-            {
-                FullName = "full name"
-            };
-            DatabaseContext.Persons.Add(person);
-            DatabaseContext.SaveChanges();
+            Person stubbedPerson = CreateStubbedPerson();
+            stubbedPerson = SavePersonToDatabase(stubbedPerson);
```
[the remark this answers is not in the document]
Comment (LBHBReynolds-Carr) [reply SocialCareCaseViewerApi.Tests/V1/Gateways/DatabaseGateway.Tests.cs]: TestHelpers.CreatePerson() now utilised

File: SocialCareCaseViewerApi.Tests/V1/Gateways/DatabaseGateway.Tests.cs [hunk 9]
```diff
@@ -606,7 +577,312 @@ public void GetWarningNotesReturnsAnExceptionIfTheWarningNoteDoesNotExist()
             act.Should().Throw<DocumentNotFoundException>()
                 .WithMessage($"No warning notes found relating to person id {request.PersonId}");
         }
+
+        [Test]
+        public void PatchWarningNoteUpdatesAnExistingRecordInTheDatabase()
+        {
+            Person stubbedPerson = CreateStubbedPerson();
```
[the remark this answers is not in the document]
Comment (LBHBReynolds-Carr) [reply SocialCareCaseViewerApi.Tests/V1/Gateways/DatabaseGateway.Tests.cs]: TestHelpers.CreatePatchWarningNoteRequest and TestHelpers.CreateWarningNote created and implemented

**Review by JohnFarrellDev — DISMISSED** [review]
Looks good to go

**Review by Renny4Real — CHANGES_REQUESTED** [review]
We don't want to merge into development until development goes into master first

**Review by JohnFarrellDev — APPROVED** [review]
Re-approve for deployment process

**Review by Renny4Real — APPROVED** [review]

**Pull request merged**